### ISSUE TITLE
Conflict with "Currency Switcher for WooCommerce"

### ISSUE THREAD

**Issue by brunocrosier** [SatelliteWP/rocket-nginx#46]
One of WP Rocket's big selling point is that it is compatible with "Currency Switcher for WooCommerce" (https://aelia.co/shop/currency-switcher-woocommerce)

Both WP Rocket and Aelia state its compatibility. 

However, when this nginx conf is enabled, the Currency Switcher plugin stops working correctly. Incorrect currencies (cached from sessions in other countries) show up as a result of Rocket-Nginx being enabled.

Would be great to have this fixed!

**Comment by maximejobin** [comment]
Thank you for posting this problem.

I'd be happy to help with that issue.  The thing is I did this project with no compensation. Buying that premium plugin for your specific need is not realistic for me.

That said, if you want to sponsor the fixing of this conflict, I would be more than happy to work on it.

**Comment by brunocrosier** [comment]
ah i see, @maximejobin - sorry i thought this was an official release by WP Rocket!

i unfortunately don't have the resources to sponsor the fixing of this conflict.. and i am sure that both WP Rocket and Aelia would benefit from this! Maybe we can ask WP Rocket and Aelia to fund it?

**Comment by maximejobin** [comment]
I understand.  This configuration is not supported by WP-Rocket.  That said, if they know what the problem is and what to help, I surely won't say no!

If someone wants to sponsor or fund it, I'm truly open to it.

**Comment by brunocrosier** [comment]
i can actually see that Aelia have written about nginx configs .. i don't know whether that would be easy to implement in your work? 

![image](https://cloud.githubusercontent.com/assets/18399089/23310592/f7763e7a-faaa-11e6-94b3-3b8de434b934.png)

link: https://aelia.freshdesk.com/support/solutions/articles/3000042591-how-to-add-dynamic-caching-to-your-site#caching-servers

**Comment by daigo75** [comment]
@brunocrosier asked me to have a look at this thread. I confirm that the solution is the one described in the knowledge base. Before serving any cached content, Nginx must take into account the value of a couple of cookies, not just the requested URL. When none of the cookies is found, then the cache should be skipped, to allow the Currency Switcher (and other plugins) to set their cookies.

The simplest flow is, basically, the following:
1. First time visitor. Cookies not set, so we don't know which content should be served. Cache should be skipped. There is no "correct default", we must identify the user first.
2. First visit is not cached. The plugins identify the customer, set the cookies, etc.
3. Second visit. The cache key must be **URL + cookies**, to ensure that different visitors will see different content, despite opening the same URL. That is:
* http://example.org/some/page + US + USD -> Content for the US, in USD.
* http://example.org/some/page + DE + EUR -> Content for Germany, in EUR.
* http://example.org/some/page + FR + EUR -> Content for France, in EUR.
* http://example.org/some/page + GB + GBP -> Content for the UK, in GBP.

And so on. This essentially creates multiple cached copies of the content for any given URL, and serves the correct content depending to each visitor.

@maximejobin If needed for testing, either Bruno or I can send you a copy of the plugin(s), so that you can check the logic on your local servers.

**Comment by maximejobin** [comment]
If cookies are set with PHP, I think Rocket-Nginx will have to be disabled because it cannot serve files statically as a cookie must be set.  If the cookie is set via Javascript, it can be made.  Can you confirm @daigo75 ?

The purpose of Rocket-Nginx is to avoid calling PHP and serve the file directly.  If some logic is needed, it might create some overhead in the Rocket-Nginx configuration file so we want to avoid that.  I will wait for the answer to see how we can deal with that.

**Comment by daigo75** [comment]
The cookies are set via PHP, as that's necessary to prepare the content to be served and, subsequently, cached. 

This won't prevent an Nginx-based solution from working. As I already explained, the key is, simply, to take the value of the cookies into account. When those cookies are not set, then the cache should be skipped altogether (i.e. the site must be hit), so that the cookies can be set and used for the next page loads. The cache skipping occurs only once, at the very first visit. 

If you would rather not skip the cache at all, not even at the first visit, then you will have to set the cookies yourself when they don't exist. This involves replicating the logic used by the Currency Switcher to perform geolocation detection and currency detection.
One of our collaborators did that using Varnish. You can find more details in our knowledge base: https://aelia.freshdesk.com/support/solutions/articles/3000042591-how-to-add-dynamic-caching-to-your-site#caching-servers

**Comment by maximejobin** [comment]
I get what you are saying. Enabling the "no cache on first visit" could be activated if the WP website owner uses your plugin.

**Comment by daigo75** [comment]
That's correct. The logic to skip the cache on first visit, as well as chaining the cookies to the cache key, should be made optional (a simple checkbox setting would do). If none of my plugins are installed, the cookies are never set, therefore there wouldn't be the need to skip the cache, nor to look for the cookies, therefore that option could be disabled by default.

**Comment by maximejobin** [comment]
This should be possible with version 3.0 as you can now include files with your own rules.

**Comment by daigo75** [comment]
@maximejobin sorry for sneaking into a closed issue. You mentioned that one can now include files with custom rules. Would you have some documentation that explains this in more detail? Thanks.

**Comment by maximejobin** [comment]
Not a problem!

Yes, I can provide more info.

You can see what can be included here:
https://github.com/SatelliteWP/rocket-nginx#configuration

I've worked on code to make Rocket-Nginx work with your plugin and, from what I remember, you'll need to use the `start.*.conf` file for setting up the `$rocket_dynamic` variable that can be used here:
https://github.com/SatelliteWP/rocket-nginx/blob/master/rocket-nginx.tmpl#L103

Please let me know if you produce the right code to make it work with your plugin. I will gladly link to it.

**Comment by maximejobin** [comment]
> For some reason, I thought it was about WP Rocket, but now I realised that it's about Nginx.

Well... first, not really. First, WP Rocket has to generate the file correctly... which it seems like they do (at first, they did not). Now, as this is not a "normal behaviour" to have the same URL produce multiple outcomes, Rocket-Nginx needs to know what to do.

The real answer is that we both (WP Rocket & Rocket-Nginx) must do something to make it work properly with the way you've coded your plugin.

I hope this helps!

**Comment by daigo75** [comment]
> Now, as this is not a "normal behaviour" to have the same URL produce multiple outcomes, Rocket-Nginx needs to know what to do.

This is something that I've been discussed quite a few times in the past, and it finds me in disagreement. The idea that one URL equals a single copy of content is not really as strict as one may think. A URL could produce the same content, but in different currencies, languages, or with different taxes. Even though, strictly speaking, the text in the page is different, from a human perspective, that would actually be the same content, just from a different point of view. 

Using a unique URL for each combination of factors is not practical. One can't have a URL like `https://example.org/path/to/some/page/<language>/country/<country>/currency/<currency>/tax/<tax>/some_other_arg/<some other value>`. Technically, it would be correct, but that would cause a stir in anyone who manages the website and who, like in 90% of cases, has no idea of what an URL does, and why. 

A workaround to these issues has been to fetch bits and pieces of the page via Ajax. That produces the same result (same URL, different content), but bots and scrapers can't see it, and the server gets pounded with a large number of calls. 

Long story short, the average site owner doesn't care about the technical aspects of a solution. They want the URLs to look pretty (i.e. no unique URLs), and pages to load fast (i.e. no Ajax). My approach, although "unorthodox", does the job. 😄 

In any case, I will take note of the information you shared. As I mentioned, I don't manage Nginx and I wouldn't be able to tinker with it, but I know quite a few people who do, and they will be happy to know about the approach you illustrated.
